base/gmp.jl

```diff
@@ -42,9 +42,16 @@ mutable struct BigInt <: Integer
     alloc::Cint
     size::Cint
     d::Ptr{Limb}
+
     function BigInt()
-        b = new(zero(Cint), zero(Cint), C_NULL)
-        MPZ.init!(b)
+        b = MPZ.init!(new())
+        finalizer(b, cglobal((:__gmpz_clear, :libgmp)))
+        return b
+    end
+
+    # n is the number of bits to allocate
+    function BigInt(::typeof(sizehint!), n)
```

Comment:
It's strange and I would rather not use this API. I would prefer a `Val{:unitit}` or `Val{:allocsize}` argument, or a dedicated type, or even `Type{Bool}`. But I wanted to avoid a keyword argument because the goal of the PR is to make a small optimization, which becomes worse than the status quo with the overhead of keyword arguments. Currently `b=BigInt()` takes 50 ns on my machine and calling `realloc(b, 200)` from GMP on it costs 65ns; with a keyword arg, `BigIng(sizehint=200)` takes about 180ns, and `BigInt()` takes 120ns !;  but `BigInt(Val{:allocsize}, 200)` takes about 50ns. Those numbers vary with the sizehint parameter (200 here), but always show that a keyword arg would be a no-go if optimization is sought after.